secrets-engines-gcp.go

```diff
@@ -0,0 +1,220 @@
+package main
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"path"
+	"strconv"
+	"text/template"
+
+	log "github.com/sirupsen/logrus"
+)
+
+type SecretsEngineGCP struct {
+	RootConfig               GcpRootConfig              `json:"root_config"`
+	OverwriteRootCredentials bool                       `json:"overwrite_root_config"`
+	ConfigLease              GcpConfigLease             `json:"config_lease"`
+	RoleSets                 map[string]gcpRoleSetEntry `json:"rolesets"`
+}
+
+type GcpRootConfig struct {
+	Credentials GcpCredentials `json:"credentials"`
+}
+
+// vault expects credentials to be a json string
+func (r GcpRootConfig) MarshalJSON() ([]byte, error) {
+	type Alias GcpRootConfig
+
+	credentials, err := json.Marshal(r.Credentials)
+	if err != nil {
+		return []byte{}, err
+	}
+
+	return json.Marshal(&struct {
+		Credentials string `json:"credentials"`
+		Alias
+	}{
+		Credentials: string(credentials),
+		Alias:       (Alias)(r),
+	})
+}
+
+type GcpCredentials struct {
+	Type                    string `json:"type"`
+	ProjectId               string `json:"project_id"`
+	PrivateKeyId            string `json:"private_key_id"`
+	PrivateKey              string `json:"private_key"`
+	ClientEmail             string `json:"client_email"`
+	ClientId                string `json:"client_id"`
+	AuthUri                 string `json:"auth_uri"`
+	TokenUri                string `json:"token_uri"`
+	AuthProviderX509CertUrl string `json:"auth_provider_x509_cert_url"`
+	ClientX509CertUrl       string `json:"client_x509_cert_url"`
+}
+
+type GcpConfigLease struct {
+	TTL    string `json:"ttl"`
+	MaxTTL string `json:"max_ttl"`
+}
+
+type gcpRoleSetEntry struct {
+	Project    string       `json:"project"`
+	SecretType string       `json:"secret_type"`
+	Bindings   []gcpBinding `json:"bindings"`
+}
+
+type gcpBinding struct {
+	Resource string   `json:"resource"`
+	Roles    []string `json:"roles"`
+}
+
+func (r gcpRoleSetEntry) BindingsAsHCL() (string, error) {
+	templateString := `{{- range .Bindings }}
+resource "{{ .Resource }}" {
+  roles = [{{- range $index, $element := .Roles }}{{- if $index }},{{ end}}"{{.}}"{{- end }}]
+}
+{{- end }}`
+	parsedTemplate, err := template.New("bindings").Parse(templateString)
+	if err != nil {
+		return "", err
+	}
+	var buf bytes.Buffer
+	if err = parsedTemplate.Execute(&buf, r); err != nil {
+		return "", err
+	}
+
+	return buf.String(), err
+}
+
+// vault expects bindings to be an hcl string
```

Comment:
Are the Vault docs wrong? it says it supports json: https://www.vaultproject.io/api-docs/secret/gcp#bindings

Reply:
It seems like it, yes.  When trying to submit bindings as a json string, it complains with `unable to parse bindings: unable to parse raw string bindings as valid HCL: At 1:1: expected: IDENT | STRING | ASSIGN | LBRACE got: LBRACK`.  When removing the brackets denoting a json array, it complains about malformed HCL: `(line 1) top-level items must have format "resource" "$resource_name"`.